source/fuzz/fuzzer_pass_add_composite_inserts.cpp

```diff
@@ -0,0 +1,226 @@
+            FindAvailableInstructions(
+                function, block, instruction_iterator,
+                [instruction_descriptor](
+                    opt::IRContext* ir_context,
+                    opt::Instruction* instruction) -> bool {
+                  // |instruction| must be a valid instruction of composite
+                  // type.
+                  if (!TransformationCompositeInsert::
+                          IsCompositeInstructionSupported(ir_context,
+                                                          instruction)) {
+                    return false;
+                  }
+                  // The instruction must be of composite type.
+                  auto instruction_type = ir_context->get_type_mgr()->GetType(
+                      instruction->type_id());
+
+                  // No components of the composite can have type
+                  // OpTypeRuntimeArray.
+                  if (ContainsRuntimeArray(*instruction_type)) {
+                    return false;
+                  }
+                  return true;
+                });
+
+        // If there are no available values, then return.
+        if (available_composites.empty()) {
+          return;
+        }
+
+        // Choose randomly one available composite value.
+        auto available_composite =
+            available_composites[GetFuzzerContext()->RandomIndex(
+                available_composites)];
+
+        // Take a random component of the chosen composite value. If the chosen
+        // component is itself a composite, then randomly decide whether to take
+        // its component and repeat. Use OpCompositeExtract to get the
+        // component.
+        bool reached_end_node = false;
+        uint32_t current_node_type_id = available_composite->type_id();
+        uint32_t one_selected_index;
+        uint32_t num_of_components;
+        std::vector<uint32_t> path_to_replaced;
+        while (!reached_end_node) {
+          num_of_components =
+              GetNumberOfComponents(GetIRContext(), current_node_type_id);
+          one_selected_index =
+              GetFuzzerContext()->GetRandomIndexForComposite(num_of_components);
+
+          // Construct a final index by appending the current index.
+          path_to_replaced.push_back(one_selected_index);
+          current_node_type_id = fuzzerutil::WalkOneCompositeTypeIndex(
+              GetIRContext(), current_node_type_id, one_selected_index);
+
+          // If the component is not a composite or if the composite is empty or
+          // if we decide not to go deeper, then end the iteration.
+          if (!fuzzerutil::IsCompositeType(
+                  GetIRContext()->get_type_mgr()->GetType(
+                      current_node_type_id))) {
+            reached_end_node = true;
+          } else if (GetNumberOfComponents(GetIRContext(),
+                                           current_node_type_id) == 0) {
+            reached_end_node = true;
```

Comment:
If `GetNumberOfComponents(GetIRContext(), current_node_type_id) == 0`, then `GetFuzzerContext()->GetRandomIndexForComposite` has already fired an assertion above.

Reply:
Also, I think it is better to get rid of `reached_end_node` and use `while(true)` and `break` here.

Reply:
Or maybe `do {...} while (...);`. Whichever you like most.